src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
+    public Map<SamPairUtil.PairOrientation, InsertSizeMetrics> getMetrics(final String groupName) throws IllegalArgumentException {
+        final Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>> intermediate = getHistogramsAndMetrics(groupName);
+        Map<SamPairUtil.PairOrientation, InsertSizeMetrics> result = new HashMap<>();
+        for(Map.Entry<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>> entry : intermediate.entrySet()){
+            result.put(entry.getKey(), entry.getValue()._2());
+        }
+        return result;
     }
 
-    public final Histogram<Integer> getHitogram(){
-        return htslibHistogram;
+    /**
+     * Utility getter for retrieving InsertSizeMetrics of a particular group, given its name and requested orientation of the pairs.
+     * If no read pairs are available in this particular group of the requested orientation, returns null.
+     * @param groupName   String representation of the group's name/id, whose InsertSizeMetrics information is requested.
+     * @param orientation Requested orientation.
+     * @return            InsertSizeMetrics of the requested group, of the requested orientation (could be null if no reads available)
+     */
+    public InsertSizeMetrics getMetricsOfOrientation(final String groupName, final SamPairUtil.PairOrientation orientation){
+        return getMetrics(groupName).get(orientation);
     }
 
-    // simply return an double array from an RDD of integers for using Apache DescriptiveStatistics
-    private static final double[] getRDDArray(final JavaRDD<Integer> rdd){
-        final List<Integer> tempList = rdd.collect();
-        final double lengthArr[] = new double[tempList.size()];
-        for(int i=0; i<lengthArr.length; ++i){
-            lengthArr[i] = tempList.get(i);
+    /**
+     * Intermediate step worker function where higher level (library, sample, all_reads) hand-rolled histograms are
+     *     constructed by merging corresponding read_group level ones.
+     * @param histogramsAtRGLevel  all hand-rolled histograms of all read groups
+     * @param toLevel              level (library, sample, all_reads) on which histograms will be built
+     * @return                     merged histogram for all available groups at the requested level
+     */
+    private static SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> levelUp(final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtRGLevel,
+                                                                                                      final MetricAccumulationLevel toLevel){
+
+        SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtHigherLevel = new TreeMap<>(groupInfoComparator);
+
+        for(final Map.Entry<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> rgStats : histogramsAtRGLevel.entrySet()){
```

Comment:
You might want to check to make sure that the input data is in fact all read group-level data - ie what would happen if someone passed a library-level histogram in in the list? I'd throw an exception in that case probably.